prov/sm2/src/sm2_ep.c

```diff
@@ -0,0 +1,987 @@
+	srx->cq = container_of(bfid, struct util_cq, cq_fid.fid);
+	ofi_atomic_inc32(&srx->cq->ref);
+	return FI_SUCCESS;
+}
+
+static void sm2_close_recv_queue(struct sm2_srx_ctx *srx,
+				 struct sm2_queue *recv_queue)
+{
+	struct fi_cq_err_entry err_entry;
+	struct sm2_rx_entry *rx_entry;
+	int ret;
+
+	while (!dlist_empty(&recv_queue->list)) {
+		dlist_pop_front(&recv_queue->list, struct sm2_rx_entry,
+				rx_entry, peer_entry);
+
+		memset(&err_entry, 0, sizeof err_entry);
+		err_entry.op_context = rx_entry->peer_entry.context;
+		err_entry.flags = rx_entry->peer_entry.flags;
+		err_entry.tag = rx_entry->peer_entry.tag;
+		err_entry.err = FI_ECANCELED;
+		err_entry.prov_errno = -FI_ECANCELED;
+		ret = srx->cq->peer_cq->owner_ops->writeerr(srx->cq->peer_cq,
+							    &err_entry);
+		if (ret)
+			FI_WARN(&sm2_prov, FI_LOG_EP_CTRL,
+				"Error writing recv entry error to rx cq\n");
+
+		ofi_freestack_push(srx->recv_fs, rx_entry);
+	}
+}
+
+static void sm2_close_unexp_queue(struct sm2_srx_ctx *srx,
+				  struct sm2_queue *unexp_queue)
+{
+	struct sm2_rx_entry *rx_entry;
+
+	while (!dlist_empty(&unexp_queue->list)) {
+		dlist_pop_front(&unexp_queue->list, struct sm2_rx_entry,
+				rx_entry, peer_entry);
+		rx_entry->peer_entry.srx->peer_ops->discard_msg(
+			&rx_entry->peer_entry);
+	}
+}
+
+static int sm2_srx_close(struct fid *fid)
+{
+	struct sm2_srx_ctx *srx;
+
+	srx = container_of(fid, struct sm2_srx_ctx, peer_srx.ep_fid.fid);
+	if (!srx)
+		return -FI_EINVAL;
+
+	sm2_close_recv_queue(srx, &srx->recv_queue);
+	sm2_close_recv_queue(srx, &srx->trecv_queue);
+
+	sm2_close_unexp_queue(srx, &srx->unexp_msg_queue);
+	sm2_close_unexp_queue(srx, &srx->unexp_tagged_queue);
+
+	ofi_atomic_dec32(&srx->cq->ref);
+	sm2_recv_fs_free(srx->recv_fs);
+	ofi_spin_destroy(&srx->lock);
+	free(srx);
+
+	return FI_SUCCESS;
+}
+
+static void cleanup_shm_resources(struct sm2_ep *ep)
+{
+	struct sm2_xfer_entry *xfer_entry;
+	struct sm2_av *av =
+		container_of(ep->util_ep.av, struct sm2_av, util_av);
+	struct sm2_mmap *map = &av->mmap;
+	struct sm2_region *self_region = sm2_mmap_ep_region(map, ep->gid);
+	bool retry = true;
+
+	/* Return all free queue entries in queue without processing them */
+return_incoming:
+	while (NULL != (xfer_entry = sm2_fifo_read(ep))) {
+		if (xfer_entry->proto == sm2_proto_return) {
+			smr_freestack_push(sm2_free_stack(sm2_mmap_ep_region(
+						   map, ep->gid)),
+					   xfer_entry);
+		} else {
+			/* TODO Tell other side that we haven't processed their
+			 * message, just returned xfer_entry */
+			sm2_fifo_write_back(ep, xfer_entry);
+		}
+	}
+
+	if (smr_freestack_isfull(sm2_free_stack(self_region))) {
+		/* TODO Set head/tail of FIFO queue to show peers we aren't
+		   accepting new entires */
+		FI_INFO(&sm2_prov, FI_LOG_EP_CTRL,
+			"All xfer_entry's acquired for map[%d], clean "
+			"shutdown\n",
+			ep->gid);
+	} else {
+		if (retry) {
+			retry = false;
+			sleep(1);
+			goto return_incoming;
+		}
```

Comment:
Still don't love this but I can't find your response (if you did respond)

Reply:
Oh I said I got rid of the `#define` and the input parameter of timeout, switched to a bool to retry once... and sleep for one second in between retries.  Its cleaner for sure, and I think we will want to re-work the cleanup logic.